### PR TITLE
Trail refactor

### PR DESCRIPTION
**Pull request by craigbranscom**
Significant refactor of Trail contract:

- Restructured notification system
- Redefined Trail definitions
- Redesigned tables
- Altered existing action signatures, added/removed signatures entirely
- New features for voting with TIP5 contracts

### CODE REVIEW

File: contracts/ratify.amend/ratify.amend.cpp [hunk 1]
```diff
@@ -9,16 +9,15 @@ ratifyamend::ratifyamend(account_name self) : contract(self), thresh_singleton(s
             _self, //publisher
             0, //initial total_voters
             0, //initial quorum_threshold
-            uint32_t(600) //expiration_length in seconds (default is 5,000,000 or ~58 days)
+            uint32_t(5000) //expiration_length in seconds (default is 5,000,000 or ~58 days)
```
Comment (ARoomWithABue) [inline contracts/ratify.amend/ratify.amend.cpp]: Is this a debug value? 
```suggestion
            uint32_t(5000000) //expiration_length in seconds (default is 5,000,000 or ~58 days)
```

File: contracts/ratify.amend/ratify.amend.cpp [hunk 2]
```diff
@@ -87,261 +85,167 @@ void ratifyamend::propose(string title, uint64_t document_id, vector<uint16_t> n
         a.title = title;
         a.new_clause_ids = new_clause_ids;
         a.new_ipfs_urls = new_ipfs_urls;
-        a.yes_count = 0;
-        a.no_count = 0;
-        a.abstain_count = 0;
+        a.yes_count = asset(0);
```
Comment (ARoomWithABue) [inline contracts/ratify.amend/ratify.amend.cpp]: Clever use of asset for calculating staked vote totals.

File: contracts/trail.service/trail.service.cpp [hunk 3]
```diff
@@ -245,4 +136,223 @@ void trail::unregballot(account_name publisher) {
...
+            print("\nTrail beginning vr search...");
+
+            uint64_t loops = 0;
+            int64_t new_no_votes = 0;
+            int64_t new_yes_votes = 0;
+            int64_t new_abs_votes = 0;
+
+            print("\nlower bound id: ", itr->receipt_id);
+
+            while(itr->vote_code == args.vote_code && loops < 10) { //loops variable to limit cpu/net expense per call
```
Comment (ARoomWithABue) [inline contracts/trail.service/trail.service.cpp]:
```suggestion
            while(itr->vote_code == args.vote_code && loops < args.loop_count) { //loops variable to limit cpu/net expense per call
```

File: contracts/trail.service/trail.definitions/traildef.voting.hpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    uint64_t proposal_id;
+    uint16_t direction;
+    uint32_t expiration;
+    account_name voter;
+};
+
+struct processvotes_args {
+    uint64_t vote_code;
+    uint64_t vote_scope;
+    uint64_t proposal_id;
```
Comment (ARoomWithABue) [inline contracts/trail.service/trail.definitions/traildef.voting.hpp]:
```suggestion
    uint64_t proposal_id;
    uint16_t loop_count;
```

File: contracts/ratify.amend/ratify.amend.cpp (hunk 1 above)
Comment (craigbranscom) [reply contracts/ratify.amend/ratify.amend.cpp]: Change value for production

**Pull request merged**